### ISSUE TITLE
REST API product update methods return 200 OK on bad input

### ISSUE THREAD

**Issue by mcphitty** [woocommerce/woocommerce#12509]
## EXPLANATION OF THE ISSUE

The REST API v1 for WooCommerce 2.6+ product update methods return HTTP 200 OK, when bad input is given. They should return HTTP 400 Bad Request, with an explanation of the error which occurred.  Returning a status of 200 OK makes debugging very difficult. When making a request to the product update methods, either PUT to /wp-json/wc/v1/products/<id>, or POST to  /wp-json/wc/v1/products/batch, if the body of the request contains a JSON object which causes a database conflict, is invalid (e.g. tries to write to a read-only field), or is malformed (e.g. missing quotation marks, curly braces, or square brackets), then the update will fail (as it should), but the response received has a status of HTTP 200 OK, and no indication is given of why the update failed.

For example, sending

{"images": [{"id": 0, src: "https://example.com/image.png", "name": "Example", "alt": "Example", "position": 0}]}

via PUT to /wp-json/wc/v1/products/<id> (assuming <id> is valid), results in a success response (status 200 OK), but the update is not done, the reason for this is (apparently) that the id of the image already exists in the database (I'm not sure if this is by design, or another issue).  The correct body should be:

{"images": [{ src: "https://example.com/image.png", "name": "Example", "alt": "Example", "position": 0}]}

This will upload a new image, and replace whatever might have been associated with the product.  Similarly, if the JSON were malformed, the response would still have a status of 200 OK, but the update would fail, since the object could not be parsed (I assume).

In either case, the response should have a status of 400 Bad Request, and the body should contain an error message explaining what went wrong, not the unmodified object or an empty object in the case of the batch update.

## STEPS TO REPRODUCE THE ISSUE

Submit bad/conflicting input or malformed JSON to the product update methods, either PUT to /wp-json/wc/v1/products/<id>, or POST to  /wp-json/wc/v1/products/batch.

**Comment by mikejolley** [comment]
What happens when only 1 prop is invalid, not all posted props?

**Comment by mcphitty** [comment]
If one prop is invalid, for example id, then the API returns a status of HTTP 200 OK, the body of the response contains the original, unmodified product, and the update is not done.  The problem is that there is no indication of why the update failed.
 
Bad Input:
 
{"images": [{"id": 1234, "src": "https://example.com/sample.png", "name": "Sample Image", "alt": "Sample Image"}]}
 
Returns HTTP 200 OK, update not done.
 
Good Input:
 
{"images": [{"src": "https://example.com/sample.png", "name": "Sample Image", "alt": "Sample Image"}]}
 
Returns HTTP 200 OK, update done - image uploaded and associated with product.

 
James F. West
Independent Software Developer
jwest@mcphitty.com
(314) 330-9736
 
 
--------- Original Message --------- Subject: Re: [woocommerce/woocommerce] REST API product update methods return 200 OK on bad input (#12509)

**Comment by claudiosanches** [comment]
@mcphitty not sure, but looks like you are testing it wrong.

> For example, sending
>
>{"images": [{"id": 0, src: "https://example.com/image.png", "name": "Example", "alt": "Example", "position": 0}]}

This is not an valid json, see the missing quotes in `src`, it should be:

```json
{"images": [{"id": 0, "src": "https://example.com/image.png", "name": "Example", "alt": "Example", "position": 0}]}
```

If you send this new json the message returned is:

```
Error: Error getting remote image https://example.com/image.png. [woocommerce_product_image_upload_error]
```

Since the image have ID 0 and don't have any `src` so will be ignored.

>Bad Input: {"images": [{"id": 1234, "src": "https://example.com/sample.png", "name": "Sample Image", "alt": "Sample Image"}]}

In this case is if the ID is a valid WordPress attachment, if not, will be ignored.
We can improve it, not a problem.

**Comment by mcphitty** [comment]
The issue was that the API returned a status of 200 OK, not an error.James F. WestIndependent Software Developerjwest@mcphitty.com(314) 330-9736


-------- Original Message --------
Subject: Re: [woocommerce/woocommerce] REST API product update methods
return 200 OK on bad input (#12509)

**Comment by mcphitty** [comment]
Also, the error message cited is because https://example.com/image.png is not a real URL.  Try the example with valid JSON (I thnk I mistyped into the ticket, it was correct in my testing), and a valid URL.  When I did that, the update failed, and the API returned a status of 200 OK, and the unmodified product.James F. WestIndependent Software Developerjwest@mcphitty.com(314) 330-9736


-------- Original Message --------
Subject: Re: [woocommerce/woocommerce] REST API product update methods
return 200 OK on bad input (#12509)
